backend/de.metas.contracts/src/main/java/de/metas/contracts/subscription/impl/SubscriptionBL.java

```diff
@@ -1018,4 +997,79 @@ public boolean isActiveTerm(@NonNull final I_C_Flatrate_Term term)
 
 		return !isCancelledOrVoided;
 	}
+
+	@Override
+	public void updatePrices(
+			@NonNull final I_C_OrderLine ol,
+			@NonNull final SOTrx soTrx,
+			final boolean updatePriceEnteredAndDiscountOnlyIfNotAlreadySet)
+	{
+		final ISubscriptionBL subscriptionBL = Services.get(ISubscriptionBL.class);
+		final IUOMConversionBL uomConversionBL = Services.get(IUOMConversionBL.class);
+		final IOrderLineBL orderLineBL = Services.get(IOrderLineBL.class);
+		final IPriceListDAO priceListDAO = Services.get(IPriceListDAO.class);
+
+		final I_C_Flatrate_Conditions flatrateConditions = ol.getC_Flatrate_Conditions();
+		final org.compiere.model.I_C_Order order = ol.getC_Order();
+
+		final PricingSystemId pricingSysytemId;
+
+		if (flatrateConditions.getM_PricingSystem_ID() > 0)
+		{
+			pricingSysytemId = PricingSystemId.ofRepoId(flatrateConditions.getM_PricingSystem_ID());
+		}
+		else
+		{
+			pricingSysytemId = PricingSystemId.ofRepoIdOrNull(order.getM_PricingSystem_ID());
+		}
+
+		final BPartnerLocationId bpLocationId = BPartnerLocationId.ofRepoId(ol.getC_BPartner_ID(), ol.getC_BPartner_Location_ID());
+		final Timestamp date = order.getDateOrdered();
+
+		final PriceListId subscriptionPLId = priceListDAO.retrievePriceListIdByPricingSyst(pricingSysytemId, bpLocationId, soTrx);
+
+		final int numberOfRuns = subscriptionBL.computeNumberOfRuns(flatrateConditions.getC_Flatrate_Transition(), date);
+
+		final Properties ctx = Env.getCtx();
+		final ProductId productId = ProductId.ofRepoIdOrNull(ol.getM_Product_ID());
+		final ProductAndCategoryId productAndCategoryId = Services.get(IProductDAO.class).retrieveProductAndCategoryIdByProductId(productId);
+		final I_C_Flatrate_Matching matching = subscriptionBL.retrieveMatching(
+				ctx,
+				ol.getC_Flatrate_Conditions_ID(),
+				productAndCategoryId,
+				ITrx.TRXNAME_None);
+
+		final Quantity qtyEntered = orderLineBL.getQtyEntered(ol);
+		final Quantity qtyOrdered = uomConversionBL.convertToProductUOM(qtyEntered, productId);
+
+		final Quantity qtyOrderedPerRun;
+		if (matching != null && matching.getQtyPerDelivery().signum() > 0)
+		{
+			final Quantity qtyPerDelivery = Quantity.of(matching.getQtyPerDelivery(), qtyOrdered.getUOM());
+			qtyOrderedPerRun = qtyPerDelivery.min(qtyOrdered);
+		}
+		else
+		{
+			qtyOrderedPerRun = qtyOrdered;
+		}
+
+		// priceQty is the qty do be delivered during one complete subscription term
+		final Quantity priceQty = qtyOrderedPerRun.multiply(numberOfRuns);
+
+		// qty ordered needs to be set because it will be used to compute the
+		// line's NetLineAmount in MOrderLine.beforeSave()
+		ol.setQtyOrdered(priceQty.toBigDecimal());
+
+		ol.setQtyEnteredInPriceUOM(priceQty.toBigDecimal());
```

Comment:
i think priceQty needs to be converted to the `ol`s C_UOM (with uomConversionBL)

Reply:
Was unsure if I should throw an error if the conversion fails, or just use the `priceQty` result. Error made more sense. Feel free to disagree.